src/support/lockedpool.cpp

```diff
@@ -97,16 +103,30 @@ void Arena::free(void *ptr)
     if (i == chunks_used.end()) {
         throw std::runtime_error("Arena: invalid or double free");
     }
-    auto freed = *i;
+    std::pair<char*, size_t> freed = *i;
     chunks_used.erase(i);
 
-    // Add space to free map, coalescing contiguous chunks
-    auto next = chunks_free.upper_bound(freed.first);
-    auto prev = (next == chunks_free.begin()) ? chunks_free.end() : std::prev(next);
-    if (prev == chunks_free.end() || !extend(prev, freed))
-        prev = chunks_free.emplace_hint(next, freed);
-    if (next != chunks_free.end() && extend(prev, *next))
+    // Coalesc freed with previous chunk
+    auto prev = chunks_free_end.find(freed.first);
```

Comment:
Why is the `chunks_free_end` map required instead of iterating one backwards from `next` as is done in the current code?

Reply:
the `cunks_free` used to be an `std::map` that was sorted by `char*. There it was possible to use `std::prev` to just find what memory address came before. I have replaced this sorted map with two `std::unordered_map`. So this change replaces one O(log(n)) lookup + std::prev() call with two O(1) lookups.